x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java

```diff
@@ -221,13 +223,25 @@ private Status buildStatus() {
     private NameID buildNameId(UserServiceAuthentication user, @Nullable SamlAuthenticationState authnState) {
         final SamlServiceProvider serviceProvider = user.getServiceProvider();
         final NameID nameID = samlFactory.object(NameID.class, NameID.DEFAULT_ELEMENT_NAME);
+        final String nameIdFormat;
         if (authnState != null && authnState.getRequestedNameidFormat() != null) {
-            nameID.setFormat(authnState.getRequestedNameidFormat());
+            nameIdFormat = authnState.getRequestedNameidFormat();
         } else {
-            nameID.setFormat(serviceProvider.getAllowedNameIdFormat() != null ? serviceProvider.getAllowedNameIdFormat() :
-                idp.getServiceProviderDefaults().nameIdFormat);
+            nameIdFormat = serviceProvider.getAllowedNameIdFormat() != null ? serviceProvider.getAllowedNameIdFormat() :
+                idp.getServiceProviderDefaults().nameIdFormat;
         }
-        // TODO: Set the value according to the format
+        nameID.setFormat(nameIdFormat);
+        nameID.setValue(getNameIdValueForFormat(nameIdFormat, user));
         return nameID;
     }
+
+    private String getNameIdValueForFormat(String format, UserServiceAuthentication user) {
+        switch (format) {
+            case TRANSIENT:
+                // See SAML 2.0 Core 8.3.8 & 1.3.4
+                return samlFactory.secureIdentifier();
+            default:
+                throw new IllegalStateException("Invalid NameID Format: " + format);
```

Comment:
```suggestion
                throw new IllegalStateException("Unsupported NameID Format: " + format);
```
I can see someone running into this in 6 months time and wondering why PERSISTENT is "invalid"